[litmus-portal]Updating to use kubernetes patch instead of get-update in subscriber for updating cm and secret

achiverram28 opened the pull request:
<!--  Thanks for sending a pull request!  -->

## Proposed changes

The ClusterRegister function in the litmus/litmus-portal/cluster-agents/subscriber/pkg/k8s
/operations.go has been modified to use kubernetes patch instead of update in subscriber for updating cm and secret


This solves the issue #3944

I have made use of the json.Marshal for the clusterData and have used it in Patch(MergePatchType) for both configMap and Secret patch.

## Types of changes

What types of changes does your code introduce to Litmus? Put an `x` in the boxes that apply
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices applies)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
- [x] I have read the [CONTRIBUTING](https://github.com/litmuschaos/litmus/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the commit for DCO to be passed.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added necessary documentation (if appropriate)

imrajdas commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -158,52 +160,42 @@ func IsClusterConfirmed() (bool, string, error) {
 
 // ClusterRegister function creates litmus-portal config map in the litmus namespace
 func ClusterRegister(clusterData map[string]string) (bool, error) {
```
We need to patch two things 
1. IS_CLUSTER_CONFIRMED
2. ACCESS_KEY

so, the func params can be be the access_key only and is_cluster_confirmed is a boolean(no need to add it in params)

```suggestion
func ClusterRegister(accessKey string) (bool, error) {
```

imrajdas commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -158,52 +160,42 @@ func IsClusterConfirmed() (bool, string, error) {
-		"SKIP_SSL_VERIFY":      clusterData["SKIP_SSL_VERIFY"],
-		"CUSTOM_TLS_CERT":      clusterData["CUSTOM_TLS_CERT"],
+	configMapPatch:= map[string]interface{}{
+		"data": clusterData,
```
only need to patch IS_CLUSTER_CONFIRMED as mentioned above

imrajdas commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -158,52 +160,42 @@ func IsClusterConfirmed() (bool, string, error) {
-		"ACCESS_KEY": clusterData["ACCESS_KEY"],
-		"CLUSTER_ID": clusterData["CLUSTER_ID"],
+	secretPatch:= map[string]interface{}{
+		"stringData":clusterData,
```
need to patch ACCESS_KEY only

imrajdas commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -158,52 +160,42 @@ func IsClusterConfirmed() (bool, string, error) {
+	}
 
-	logrus.Info(AgentSecretName + " has been updated")
+	logrus.Info(AgentSecretName+" has been updated")
```
```suggestion
	logrus.Infof("%s has been patched", AgentSecretName)
```

imrajdas commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -158,52 +160,42 @@ func IsClusterConfirmed() (bool, string, error) {
 	}
 
-	logrus.Info(AgentConfigName + " has been updated")
+	logrus.Info(AgentConfigName+" has been updated")
```
```suggestion
	logrus.Infof("%s has been patched", AgentConfigName)
```

achiverram28 replied:
Sure Ill makes these changes and update the PR

imrajdas reviewed (changes requested):
Thanks for this PR. Gave few comments, PTAL

imrajdas commented:
cc: @gdsoumya @amityt for reviews

achiverram28 commented:
@imrajdas I have updated the code and updated the pr . Can the maintainers have a look at it

imrajdas commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -157,53 +159,48 @@ func IsClusterConfirmed() (bool, string, error) {
 // ClusterRegister function creates litmus-portal config map in the litmus namespace
-func ClusterRegister(clusterData map[string]string) (bool, error) {
-	clientset, err := GetGenericK8sClient()
+func ClusterRegister(accessKey string) (bool, error) {
```
Check where ClusterRegister() is being called and update the params there

gdsoumya commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -157,53 +159,48 @@ func IsClusterConfirmed() (bool, string, error) {
-		"VERSION":              clusterData["VERSION"],
-		"SKIP_SSL_VERIFY":      clusterData["SKIP_SSL_VERIFY"],
-		"CUSTOM_TLS_CERT":      clusterData["CUSTOM_TLS_CERT"],
+	isClusterConfirmed, _, err := IsClusterConfirmed()
+	if err != nil {
+		return false, err
```
No need to call it again, pass it as param as suggested by raj.

gdsoumya commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -157,53 +159,48 @@ func IsClusterConfirmed() (bool, string, error) {
-		},
-	}, metav1.UpdateOptions{})
+	configMapPatch:= map[string]bool {
+		"data": isClusterConfirmed,
```
Did you test if this works? You need to specify the key-value mapping for the patch, here you just put the bool value directly under `data`

amityt replied:
@achiverram28  You can try something like this
		
            updatedConfigMap := corev1.ConfigMap{
			Data: map[string]string{
				"<key>": "<value>",
			},
		}

		configMapPatch, err := json.Marshal(updatedConfigMap)
		if err != nil {
			return fmt.Errorf("failed to marshal patched config, err: %v", err)
		}

gdsoumya commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -157,53 +159,48 @@ func IsClusterConfirmed() (bool, string, error) {
+	logrus.Info("%s has been patched",AgentConfigName)
+
+	secretPatch:= map[string]string{
+		"stringData": accessKey,
```
same here need to specify the map

gdsoumya reviewed (changes requested):
Please take a look at how patch works and whats the expected inputs for it and test the code locally to make sure it works.

amityt commented on litmus-portal/cluster-agents/subscriber/pkg/k8s/operations.go:
```diff
@@ -157,53 +159,48 @@ func IsClusterConfirmed() (bool, string, error) {
-		"VERSION":              clusterData["VERSION"],
-		"SKIP_SSL_VERIFY":      clusterData["SKIP_SSL_VERIFY"],
-		"CUSTOM_TLS_CERT":      clusterData["CUSTOM_TLS_CERT"],
+	isClusterConfirmed, _, err := IsClusterConfirmed()
```
As mentioned by Shomo, this function is not needed here. Instead, you can directly set the value to `true` since we are already updating the map here: https://github.com/litmuschaos/litmus/blob/cf86f7e308695a2bdfbeeb585f3ab809b81dcdcd/litmus-portal/cluster-agents/subscriber/subscriber.go#L115

amityt reviewed (commented):
Please make the requested changes @achiverram28

achiverram28 commented:
Yes , I will work on all of these changes and update the pr . Thanks for the guidance

achiverram28 commented:
I have made changes by updating the params of the ```ClusterRegister()``` in ```subscriber.go``` @imrajdas .
I have removed the function call inside and studied the ```subscriber.go ``` and passed it as a parameter to ```ClusterRegister()``` . So I have passed ```accessKey``` and ```isClusterConfirmed``` as arguments in ```subscriber.go``` @imrajdas @gdsoumya @amityt .
Sorry for the mistake during ```ConfigMap``` and  ```Secret``` creation . As per the guidance of @amityt and @gdsoumya I properly studied this  usage and input for those and modified by adding maps . 
Hope this solves all the requested changes . If any more changes I am open for it.
